torchtune/modules/optim/cpu_offload.py

```diff
@@ -0,0 +1,77 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from torch.optim.optimizer import ParamsT
+from torchtune.config._utils import _get_component_from_path
+
+try:
+    from torchao.prototype.low_bit_optim import (
+        CPUOffloadOptimizer as _CPUOffloadOptimizer,
+    )
+
+    _CPU_OFFLOAD_OPTIM_AVAILABLE = True
+
+    class CPUOffloadOptimizer(_CPUOffloadOptimizer):
+        """Offload optimizer to CPU for single-GPU training. This will reduce GPU memory by the size of optimizer
+        state. Optimizer step will be done on CPU.
+
+        Args:
+            params (ParamsT): a list of parameters or parameter groups.
+            optimizer_class (str): name of the base optimizer.
+            offload_gradients (bool): free GPU gradients once they are moved to CPU. Setting this to True will further
+                reduce GPU memory by the size of gradients. This is not compatible with gradient accumulation.
+            **kwargs: other keyword arguments to be passed to the base optimizer e.g. `lr`, `weight_decay`.
+
+        Example:
+            >>> from torchtune.modules.optim import CPUOffloadOptimizer
+            >>> optimizer = CPUOffloadOptimizer(
+            ...    params=model.parameters(),
+            ...    optimizer_class='torch.optim.AdamW',
+            ...    offload_gradients=False,
+            ...    lr=2e-5
+            ... )
+
+        To use this optimizer from the command line, you can write
+            optimizer._component_=torchtune.modules.optim.CPUOffloadOptimizer
+            optimizer.optimizer_class=torch.optim.AdamW
+
+        NOTE: This is a light wrapper around `torchao.prototype.low_bit_optim.CPUOffloadOptimizer` that parses
+        optimizer class string into a concrete optimizer class. See the documentation here for caveats
+        https://github.com/pytorch/ao/tree/main/torchao/prototype/low_bit_optim#optimizer-cpu-offload
+        """
+
+        def __init__(
+            self,
+            params: ParamsT,
+            optimizer_class: str,
+            *,
+            offload_gradients: bool = False,
+            **kwargs,
+        ) -> None:
+            super().__init__(
+                params=params,
+                optimizer_class=_get_component_from_path(optimizer_class),
```

Comment:
As if there aren't enough cooks in the kitchen already on this discussion.. one option could be to just do this all in the recipe:

```
if hasattr(cfg.optimizer, "optimizer_class"):
	torchao_version, is_nightly = _get_torchao_version()
	if not (**some condition on ao version**):
		raise ValueError("CPU offload optimizer not supported")
		cfg.optimizer.optimizer_class = _get_component_from_path(optimizer_class)
		self._optimizer = instantiate(cfg.optimizer)
else:
	# normal optimizer instantiation
```

Not 100% sure this will work btw, please lmk if it's a dumb suggestion. But the idea is to hopefully avoid the need for a thin wrapper class just to do str -> type. 

To weigh in on the other parts of the conversation though, I do think your usage of `_get_component_from_path` is the way to go here. If we're gonna use it like this we should make it a public API though (but that doesn't have to be your problem)

Reply:
Something like you suggested should work, but I want to avoid adding extra optimizer parsing logic to ALL recipes :sweat_smile:.

Reply:
Yeah that's understandable. Not that this is a sound argument, but there is precedent for doing something like this in torchtune already: see our [ConcatDataset](https://github.com/pytorch/torchtune/blob/e568b671bf4dd043649611c3d0905967257ccac2/recipes/full_finetune_distributed.py#L448-L457), which is pretty much the same in every recipe. I think this is kind of what the `_setup_*` methods are for anyways.. in an ideal world we get that nice one-line instantiation of every component, but in practice without recursive instantiation that's unlikely to happen (I am also happy to revisit that topic but iiuc it won't solve this particular case anyways)